ethcore/light/src/client/snapshot/chain.rs

```diff
@@ -0,0 +1,107 @@
+// Copyright 2015-2018 Parity Technologies (UK) Ltd.
+// This file is part of Parity.
+
+// Parity is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Parity is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Parity.  If not, see <http://www.gnu.org/licenses/>.
+
+//! `RestorationTargetChain` impl for `HeaderChain`.
+
+
+use super::super::HeaderChain;
+use client::header_chain::PendingChanges;
+use ethcore::engines::EpochTransition;
+use ethcore::ids::BlockId;
+use ethcore::header::{Header, BlockNumber};
+use ethcore::snapshot::RestorationTargetChain;
+use ethereum_types::{H256, U256};
+use parking_lot::RwLock;
+use kvdb::DBTransaction;
+use ethcore::receipt::Receipt;
+use ethcore::views::BlockView;
+
+/// Wrapper for `HeaderChain` along with pending changes.
+pub struct LightChain {
+	chain: HeaderChain,
+	pending: RwLock<Option<PendingChanges>>
+}
+
+impl LightChain {
+	/// Create a wrapper for `HeaderChain` implementing `RestorationTargetChain`
+	pub fn new(chain: HeaderChain) -> Self {
+		LightChain {
+			chain: chain,
+			pending: RwLock::new(None)
+		}
+	}
+
+	/// Get a reference to the underlying chain
+	pub fn chain(&self) -> &HeaderChain {
+		&self.chain
+	}
+}
+
+impl RestorationTargetChain for LightChain {
+	fn genesis_hash(&self) -> H256 {
+		self.chain.genesis_hash()
+	}
+
+	fn genesis_header(&self) -> Header {
+		self.chain.genesis_header().decode().expect("genesis header is always decodable; qed")
+	}
+
+	fn block_hash(&self, index: BlockNumber) -> Option<H256> {
+		self.chain.block_hash(BlockId::Number(index))
+	}
+
+	fn block_header_data(&self, hash: &H256) -> Option<Header> {
+		self.chain.block_header(BlockId::Hash(hash.clone())).and_then(|h| h.decode().ok())
+	}
+
+	fn add_child(&self, _batch: &mut DBTransaction, _block_hash: H256, _child_hash: H256) {
+		// TODO
+	}
+
+	fn insert_epoch_transition(
+		&self, batch: &mut DBTransaction,
+		header: Header, transition: EpochTransition
+	) {
+		let result = if header.number() == 0 {
+			let td = self.chain.genesis_header().difficulty();
+			self.chain.insert_with_td(batch, header, td, Some(transition.proof))
+		} else {
+			self.chain.insert(batch, header, Some(transition.proof))
+		};
+		match result {
+			Ok(pending) => *self.pending.write() = Some(pending),
+			// TODO: panic? abort snapshot sync?
```

Comment:
Why not return error upstream and have?

```rust
fn insert_epoch_transition(… …) -> Result<…> {
  if header.number() == 0 {
    let td = self.chain.genesis_header().difficulty();
    self.chain.insert_with_td(batch, header, td, Some(transition.proof))
  } else {
    self.chain.insert(batch, header, Some(transition.proof))
  }
}
```

**EDIT** I see: the new trait wraps previous code that doesn't use `Return` much. Is it worthwhile to change that perhaps, so that new code can use more idiomatic error handling?